Pending identity inserts

kustosz opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: <https://github.com/Recni/rust-app-template/blob/master/CONTRIBUTING.md>

The contributors guide includes instructions for running rustfmt and building the
documentation.
-->

## Motivation
Part of the re-arch project, this code corresponds to processes a. and b. as defined in #258 . Process C is to be cut out based on #270.


## Solution

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Updated the changelog

<!-- This template is based on https://github.com/tokio-rs/tokio/blob/tokio-1.13.0/.github/PULL_REQUEST_TEMPLATE.md and https://github.com/gakonst/ethers-rs/blob/0.5.3/.github/PULL_REQUEST_TEMPLATE.md -->

recmo commented on src/app.rs:
```diff
@@ -476,6 +479,144 @@ impl App {
+        let mut listener = self.event_bus.subscribe();
+        tokio::spawn(async move {
+            let mut transaction_in_progress = false;
+            let (tx, mut rx) = tokio::sync::mpsc::channel(1);
```
These names are not descriptive. I'm also not sure if `mspc` is the appropriate channel type.

recmo commented on src/app.rs:
```diff
@@ -476,6 +479,144 @@ impl App {
+        tokio::spawn(async move {
+            let mut transaction_in_progress = false;
+            let (tx, mut rx) = tokio::sync::mpsc::channel(1);
+            let mut work_queue: VecDeque<(usize, Hash)> = VecDeque::new();
```
Isn't this redundant with the queue in the `event_bus` channel and the new database table?

recmo commented on src/app.rs:
```diff
@@ -476,6 +479,144 @@ impl App {
+                    else => {}
+                }
+            }
+        });
```
Why not

```
while let Some(event) = listener.recv().await {
 app.commit_identity(...
}
```
?

recmo replied:
Or even simpler, just use the database as work queue. And poll the database in a loop.

To make it reactive we can have it stop when there is no more work and use an event to start the loop. This can be a simple [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html) and doesn't have to carry state like the event_bus.

recmo reviewed (changes requested):
I'm a bit concerned with the work queue being stored in three different places.

1. In the database.
2. In the event bus channel.
3. In the work_queue VecDeque

I think we should stick to using the database exclusively. That should simplify things and avoid potential issues with keeping state in sync.

kustosz commented on src/timed_read_progress_lock.rs:
```diff
@@ -0,0 +1,183 @@
+///
+/// Can be used to specify a critical modification section, where certain
+/// computations should be exclusive, even if the structure can still be read
+/// (possibly getting stale results), narrowing the read-excluding section.
```
I don't think so, at least I'm not aware of this. It's just a primitive I needed but couldn't find :)

kustosz commented on src/app.rs:
```diff
@@ -476,6 +479,144 @@ impl App {
+        tree.set(identity_index, commitment);
+
+        // Downgrade write lock to progress lock
+        let tree = tree.downgrade_to_progress();
```
This looks like a serious sanity check, so I don't think it's a good idea to allow something to run concurrently from a broken state. Or if we're OK with another job picking up a broken state, maybe we should remove this check altogether (and allow the next insert to fail).

kustosz commented:
@recmo you're right, it was better to just poll the DB. Also I ended up holding on to the thread handle, so it can be supervised / restarted in the future. I think I addressed all of your feedback, either by changing code or following up in the comments.

recmo commented on src/database.rs:
```diff
@@ -125,6 +125,82 @@ impl Database {
+            )
+            .await?;
+        Ok(())
+    }
```
Stylistic nitpick: I think it looks better with an intermediate variable:
```rust
        let query = sqlx::query(
            r#"UPDATE pending_identities
                    SET mined_in_block = $1, assigned_leaf_idx = $2
                    WHERE group_id = $3 AND commitment = $4;"#,
        )
        .bind(block_number as i64)
        .bind(tree_idx as i64)
        .bind(group_id as i64)
        .bind(commitment);
        self.pool.execute(query).await?;
        Ok(())
```
This reduces the amount of identation and lines just containing `)`.

recmo commented on src/database.rs:
```diff
@@ -125,6 +125,82 @@ impl Database {
+                           LIMIT 1;"#,
+            ))
+            .await?;
+        Ok(row.map(|row| (usize::try_from(row.get::<i64, _>(0)).unwrap(), row.get(1))))
```
It's a bit annoying that there is no `usize` impl, but understandable since some databases (i.e. postgres) don't have a `u64` type.

FWIW, this is a bit more concise `row.get::<i64, _>(0).try_into().unwrap()`

recmo commented on src/identity_committer.rs:
```diff
@@ -0,0 +1,204 @@
+        // which is impossible, since this is the only use, and this method takes
+        // ownership, or the channel is closed, which means the committer thread is
+        // already dead.
+        self.shutdown_sender.send(()).await.ok();
```
`.ok()` just returns an `Option<T>` containing the value, if any. If we want to dispose of the result without triggering a panic or the `must_use` lint the pattern is:

```rust
let _ = self.shutdown_sender.send(()).await;
```

recmo commented on src/identity_committer.rs:
```diff
@@ -0,0 +1,204 @@
+    pub async fn start(&self) {
+        let mut instance = self.instance.write().await;
+        if instance.is_some() {
+            info!("Identity committer already running");
```
This should at least be a warning, and it is arguably an error if this happens.

kustosz replied:
I'll bump it to a warning, I don't think it is an error. In the future, we may want to restart the committer and this behavior allows the first call that spots it's dead to do so safely – if someone else already restarted, it'll just be a no-op.

recmo commented on src/identity_committer.rs:
```diff
@@ -0,0 +1,204 @@
+        let database = self.database.clone();
+        let tree_state = self.tree_state.clone();
+        let contracts = self.contracts.clone();
+        let handle = tokio::spawn(async move {
```
This basically creates a closure taking the variables cloned above. In the future when this gets more complex we probably want a explicit struct for the context like so:

```rust
struct Task {
  database: // ...
  // ..
}

impl Task {
   async fn run(mut self) { ... }
}
```

And then launch it by 

```
let task = Task { ... };
let handler = tokio::spawn(task.run());
```

recmo commented on src/identity_committer.rs:
```diff
@@ -0,0 +1,204 @@
+            })?;
+
+        // Immediately write the tree to storage, before anyone else can
+        // TODO: Store tree in database
```
I think this comment is copied over, but no longer relevant? At any rate the TODO is captured in one of the open issues, so we can remove it here.

recmo commented on src/identity_committer.rs:
```diff
@@ -0,0 +1,204 @@
+            block, identity_index
+        );
+        database
+            .mark_identity_inserted(group_id, &commitment, block.as_usize(), identity_index)
```
This works as long as there is a single `identity_comitter`.
If there is more than one we need to transactionally fetch a batch and flag them as 'in progress'. That is out of scope for this PR.
But we should at least add a warning in the `IdentityCommitter`  documentation (e.g. a `///` comment).

kustosz commented on src/app.rs:
```diff
@@ -71,14 +72,32 @@ pub struct Options {
+pub struct TreeState {
+    pub next_leaf:   AtomicUsize,
+    pub merkle_tree: TimedReadProgressLock<PoseidonTree>,
+}
```
Good point, strange how it was allowed to be changed independently.
